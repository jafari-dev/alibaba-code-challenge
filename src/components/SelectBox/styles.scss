.selectbox {
  width: 200px;
  padding: 12px 8px;
  font-family: "Nunito Sans", sans-serif;
  font-size: 16px;
  font-weight: 600;
  border: none;
  outline: none;
  box-shadow: 0 0 4px var(--shadow-color);
  background-color: var(--secondary-background-color);
  color: var(--text-color);
  cursor: pointer;

  option {
    font-weight: 600;
  }
}
