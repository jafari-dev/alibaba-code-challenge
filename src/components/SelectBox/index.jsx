import "./styles.scss";
import { memo, useCallback } from "react";


function SelectBox({options, onChange, placeholder}) {
  const changeSelectedOption = useCallback((event) => {
    const value = event.target.value;
    onChange(value);
  }, [onChange]);

  return (
    <select defaultValue="" className="selectbox" data-testid="selectbox" name="select" onChange={changeSelectedOption}>
      <option data-testid="option" disabled value="">{placeholder}</option>
      {options.map((option) => (
        <option data-testid="option" key={option} value={option}>
          {option}
        </option>
      ))}
    </select>
  )
}

export default memo(SelectBox);
