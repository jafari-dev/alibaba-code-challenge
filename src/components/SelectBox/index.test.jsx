/* eslint-disable jest/no-conditional-expect */
/* eslint-disable testing-library/prefer-screen-queries */
import { fireEvent, render } from "@testing-library/react";
import { faker } from "@faker-js/faker";
import SelectBox from ".";

beforeAll(() => {
  faker.seed(1);
});

describe("SelectBox component", () => {
  test("matches the snapshots.", () => {
    const options = Array.from(
      {
        length: faker.datatype.number({ min: 3, max: 10 }),
      },
      () => faker.random.word()
    );

    const { container } = render(
      <SelectBox
        placeholder={faker.random.word()}
        options={options}
        onChange={jest.fn()}
      />
    );

    expect(container).toMatchSnapshot();
  });

  test("sets an empty string as the default value of selectbox.", () => {
    const options = Array.from(
      {
        length: faker.datatype.number({ min: 3, max: 10 }),
      },
      () => faker.random.word()
    );

    const { getByTestId } = render(
      <SelectBox
        placeholder={faker.random.word()}
        options={options}
        onChange={jest.fn()}
      />
    );

    expect(getByTestId("selectbox")).toHaveValue("");
  });

  test("calls the provided `onChange` by changing the selectbox value.", () => {
    const onChange = jest.fn();
    const options = Array.from(
      {
        length: faker.datatype.number({ min: 3, max: 10 }),
      },
      () => faker.random.word()
    );

    const { getByTestId } = render(
      <SelectBox
        placeholder={faker.random.word()}
        options={options}
        onChange={onChange}
      />
    );

    expect(onChange).not.toBeCalled();

    fireEvent.change(
      getByTestId("selectbox", {
        target: { value: faker.helpers.arrayElement(options) },
      })
    );
  });

  test("contains a disabled option tag with the provided 'placeholder' prop.", () => {
    const options = Array.from(
      {
        length: faker.datatype.number({ min: 3, max: 10 }),
      },
      () => faker.random.word()
    );
    const placeholder = faker.random.word();

    const { getAllByTestId } = render(
      <SelectBox
        placeholder={placeholder}
        options={options}
        onChange={jest.fn()}
      />
    );

    expect(getAllByTestId("option")[0]).toHaveValue("");
    expect(getAllByTestId("option")[0]).toBeDisabled();
    expect(getAllByTestId("option")[0]).toHaveTextContent(placeholder);
  });

  test("contains a list of option tags that they have the", () => {
    const onChange = jest.fn();
    const options = Array.from(
      {
        length: faker.datatype.number({ min: 3, max: 10 }),
      },
      () => faker.random.word()
    );

    const { getAllByTestId } = render(
      <SelectBox
        placeholder={faker.random.word()}
        options={options}
        onChange={onChange}
      />
    );

    const allOptions = [...getAllByTestId("option")];

    // The first ignored option is the default option (previous test)
    const [, ...normalOptions] = allOptions;

    expect(allOptions).toHaveLength(options.length + 1);

    normalOptions.forEach((option, index) => {
      expect(option).toHaveTextContent(options[index]);
    });
  });
});
