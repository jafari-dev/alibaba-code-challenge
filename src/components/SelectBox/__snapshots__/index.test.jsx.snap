// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectBox component matches the snapshots. 1`] = `
<div>
  <select
    class="selectbox"
    data-testid="selectbox"
    name="select"
  >
    <option
      data-testid="option"
      disabled=""
      selected=""
      value=""
    >
      group
    </option>
    <option
      data-testid="option"
      value="remove"
    >
      remove
    </option>
    <option
      data-testid="option"
      value="yuck"
    >
      yuck
    </option>
    <option
      data-testid="option"
      value="Kansas"
    >
      Kansas
    </option>
    <option
      data-testid="option"
      value="cloak"
    >
      cloak
    </option>
    <option
      data-testid="option"
      value="Electronic"
    >
      Electronic
    </option>
    <option
      data-testid="option"
      value="collaboration"
    >
      collaboration
    </option>
  </select>
</div>
`;
